### PR TITLE
refactor: parachain template

### PR DESCRIPTION
**Pull request by Daanvdplas**
Update parachain template pallet based on polkadot sdk docs

@kianenigma

### CODE REVIEW

**Review by kianenigma — APPROVED** [review]
Great! 

CC @gupnik @rzadp let's agree on all changes here, then mirror it in all 3 pallet templates?

File: templates/parachain/pallets/template/src/lib.rs [hunk 1]
```diff
@@ -62,9 +62,10 @@ pub mod pallet {
 	#[pallet::hooks]
 	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
 
-	// Dispatchable functions allows users to interact with the pallet and invoke state changes.
-	// These functions materialize as "extrinsics", which are often compared to transactions.
-	// Dispatchable functions must be annotated with a weight and must return a DispatchResult.
+	/// Dispatchable functions allows users to interact with the pallet and invoke state changes.
+	/// These functions materialize as "extrinsics", which are often compared to transactions.
+	/// Dispatchable functions must be annotated with a weight and must return a DispatchResult.
+	/// https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/guides/your_first_pallet/index.html#dispatchables
```
Comment (kianenigma) [inline templates/parachain/pallets/template/src/lib.rs:68]: Also link to the `pallet::storage` macro doc, they are pretty good.

File: templates/parachain/pallets/template/src/mock.rs [hunk 2]
```diff
@@ -1,50 +1,21 @@
...
+use frame_system::{GenesisConfig, mocking::MockBlock};
+use sp_runtime::{traits::ConstU64, BuildStorage};
 
 // Configure a mock runtime to test the pallet.
 frame_support::construct_runtime!(
-	pub enum Test
-	{
-		System: frame_system::{Pallet, Call, Config<T>, Storage, Event<T>},
-		TemplateModule: crate::{Pallet, Call, Storage, Event<T>},
+	pub enum Test {
```
Comment (kianenigma) [inline templates/parachain/pallets/template/src/mock.rs:7]:
```suggestion
	pub struct Test {
```

File: templates/parachain/pallets/template/src/lib.rs [hunk 3]
```diff
@@ -16,6 +13,8 @@ pub mod weights;
 #[cfg(feature = "runtime-benchmarks")]
 mod benchmarking;
 
+// https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/frame_runtime/index.html
+// https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/guides/your_first_pallet/index.html
```
Comment (kianenigma) [inline templates/parachain/pallets/template/src/lib.rs:17]:
```suggestion
// https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/guides/your_first_pallet/index.html
// To see a full list of `pallet` macros and their use cases, see: 
// https://paritytech.github.io/polkadot-sdk/master/pallet_example_kitchensink/index.html
// https://paritytech.github.io/polkadot-sdk/master/frame_support/pallet_macros/index.html
```

File: templates/parachain/pallets/template/src/lib.rs (hunk 1 above)
Comment (Daanvdplas) [reply templates/parachain/pallets/template/src/lib.rs:72]: See line 43

File: templates/parachain/pallets/template/src/lib.rs (hunk 3 above)
Comment (Daanvdplas) [reply templates/parachain/pallets/template/src/lib.rs:17]: Added

File: templates/parachain/pallets/template/src/mock.rs (hunk 2 above)
Comment (kianenigma) [reply templates/parachain/pallets/template/src/mock.rs]: @gupnik we can move all of this to the new syntax as well right?

Comment (gupnik) [reply templates/parachain/pallets/template/src/mock.rs]: > @gupnik we can move all of this to the new syntax as well right?

Yes! @Daanvdplas Do you want to give that a try? https://github.com/paritytech/polkadot-sdk/pull/1378 should help.

Comment (Daanvdplas) [reply templates/parachain/pallets/template/src/mock.rs]: Will finish this weekend

Comment (Daanvdplas) [reply templates/parachain/pallets/template/src/mock.rs]: It also compiled without `experimental` feature, didn't run a local network with it though, added it because I saw it being used in other places.

**Comment by Daanvdplas** [comment]
>CC @gupnik @rzadp let's agree on all changes here, then mirror it in all 3 pallet templates?

With pop cli we are working on a feature `pop new pallet` & `pop add pallet`. This could mean that the templates offered in the polkadot sdk (i.e.  minimal, solo & parachain) won't need the pallet template included. In stead, a separate pallet crate can be created that can be used by all three.

File: templates/parachain/runtime/src/lib.rs [hunk 4]
```diff
@@ -232,43 +231,69 @@ pub fn native_version() -> NativeVersion {
...
+	#[runtime::pallet_index(31)]
+	pub type PolkadotXcm = pallet_xcm;
+	#[runtime::pallet_index(32)]
+	pub type CumulusXcm = cumulus_pallet_xcm;
+	#[runtime::pallet_index(33)]
+	pub type MessageQueue = pallet_message_queue;
+	#[runtime::pallet_index(50)]
+
+	// Template
+	pub type TemplatePallet = pallet_parachain_template;
```
Comment (gupnik) [inline templates/parachain/runtime/src/lib.rs:293]: Missing pallet index. Should throw an error when you merge the latest master.

File: templates/parachain/runtime/src/lib.rs [hunk 5]
```diff
@@ -232,43 +231,69 @@ pub fn native_version() -> NativeVersion {
...
+	pub type XcmpQueue = cumulus_pallet_xcmp_queue;
+	#[runtime::pallet_index(31)]
+	pub type PolkadotXcm = pallet_xcm;
+	#[runtime::pallet_index(32)]
+	pub type CumulusXcm = cumulus_pallet_xcm;
+	#[runtime::pallet_index(33)]
+	pub type MessageQueue = pallet_message_queue;
+	#[runtime::pallet_index(50)]
+
+	// Template
```
Comment (gupnik) [inline templates/parachain/runtime/src/lib.rs:292]: Some formatting issue it seems.

File: templates/parachain/runtime/src/lib.rs [hunk 6]
```diff
@@ -7,6 +7,8 @@
 include!(concat!(env!("OUT_DIR"), "/wasm_binary.rs"));
 
 pub mod apis;
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarks;
```
Comment (gupnik) [inline templates/parachain/runtime/src/lib.rs:11]: Why is this removed?

**Review by gupnik — APPROVED** [review]

File: templates/parachain/pallets/template/src/mock.rs (hunk 2 above)
Comment (gupnik) [reply templates/parachain/pallets/template/src/mock.rs]: Yeah, `runtime` macro is out of `experimental` now.

File: templates/parachain/runtime/src/lib.rs (hunk 6 above)
Comment (Daanvdplas) [reply templates/parachain/runtime/src/lib.rs:11]: Moved it up the top with the rest of the module declarations

File: templates/parachain/runtime/src/lib.rs (hunk 5 above)
Comment (Daanvdplas) [reply templates/parachain/runtime/src/lib.rs:292]: Wdym?

File: templates/parachain/pallets/template/Cargo.toml [hunk 7]
```diff
@@ -22,7 +22,7 @@ scale-info = { version = "2.11.1", default-features = false, features = [
 
 # frame deps
 frame-benchmarking = { path = "../../../../substrate/frame/benchmarking", default-features = false, optional = true }
-frame-support = { path = "../../../../substrate/frame/support", default-features = false }
+frame-support = { path = "../../../../substrate/frame/support", default-features = false, features = ["experimental"] }
```
Comment (bkchr) [inline templates/parachain/pallets/template/Cargo.toml:25]: We should not add experimental stuff to the template. The point of it being that it is still experimental and subject to change.

Comment (Daanvdplas) [reply templates/parachain/pallets/template/Cargo.toml:25]: So revert the changes regarding construct runtime v2? @gupnik

Comment (gupnik) [reply templates/parachain/pallets/template/Cargo.toml:25]: > So revert the changes regarding construct runtime v2? @gupnik

It's not experimental anymore. You should be able to use it without that flag @Daanvdplas.

File: templates/parachain/runtime/src/lib.rs (hunk 5 above)
Comment (gupnik) [reply templates/parachain/runtime/src/lib.rs:292]: It would be better to have
```rust
// Template
#[runtime::pallet_index(50)
pub type Template = ..
```

File: templates/parachain/pallets/template/Cargo.toml (hunk 7 above)
Comment (bkchr) [reply templates/parachain/pallets/template/Cargo.toml:25]: Then we should not enable the experimental feature :P

Comment (kianenigma) [inline templates/parachain/pallets/template/Cargo.toml:25]:
```suggestion
frame-support = { path = "../../../../substrate/frame/support", default-features = false }
```

Comment (Daanvdplas) [reply templates/parachain/pallets/template/Cargo.toml]: Apologies, I thought seeing it being used in the kitchen sink runtime but I think I had a bad dream. Thanks Kian

**Comment by Daanvdplas** [comment]
Should be good to go now

**Review by skunert — APPROVED** [review]
Looks nice!
